### PR TITLE
[RAPTOR-6370] more robust detection of text vs categorical

### PR DESCRIPTION
**Pull request by scottp-ml**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
This catches cases where categorical values have whitespace in them.  In such cases they were originally being treated as text, while in DR they are categorical.  Now the check includes the number and percent of unique values as well.

### CODE REVIEW

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 1]
```diff
@@ -236,7 +236,8 @@ def is_text(x):
         """
         Decide if a pandas series is text, using a very simple heuristic:
         1. Count the number of elements in the series that contain 1 or more whitespace character
-        2. If >75% of the elements have whitespace, the Series is text
+        2. If >75% of the elements have whitespace, the Series is text and there are more than 60 unique values or
+            more than 5% of values are unique
```
Comment (drenfr01) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:240]: @scottp-ml I'm not sure I understand this comment. I believe it's saying:

"The series is determined to be text IF >75% of the elements have whitespace and (there are more than 60 unique values or more than 5% of values are unique)"

Is that correct?

**Review by drenfr01 — APPROVED** [review]
Feel free to clean up that comment in another PR. Otherwise LGTM

**Comment by 4dahalibut** [comment]
can you cull like all but 100 rows of the csv for storage reasons

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py (hunk 1 above)
Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:240]: yes please rewrite for clarity

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 2]
```diff
@@ -246,13 +247,19 @@ def is_text(x):
...
             pd.api.types.is_string_dtype(x)
             and pd.api.types.infer_dtype(x) != "boolean"
             and pd.api.types.infer_dtype(x) != "bytes"
         ):
             pct_rows_with_whitespace = (x.str.count(r"\s") > 0).sum() / x.shape[0]
-            return pct_rows_with_whitespace > 0.75
+            unique = x.nunique()
+            pct_unique_values = unique / x.shape[0]
+            return pct_rows_with_whitespace >= MIN_WHITESPACE_ROWS and (
+                pct_unique_values >= MIN_UNIQUE_VALUES or unique >= 60
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:261]: idk if we want unique >= 60, what about 100 class multiclass?

Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:261]: @mikhail-dr how do you feel about this heuristic?

**Review by 4dahalibut — APPROVED** [review]

**Review by 4dahalibut — COMMENTED** [review]
some questions

**Review by 4dahalibut — APPROVED** [review]

**Pull request merged**